### ISSUE TITLE
Project dependencies may have API risk issues

### ISSUE THREAD

**Issue by PyDR-Tool** [baidu-research/NCRF#67]
Hi, In **NCRF**, inappropriate dependency versioning constraints can cause risks.

Below are the dependencies and version constraints that the project is using

```
numpy==1.14.3
scipy==1.0.1
torchvision==0.2.0
pillow==8.1.1
scikit-image==0.13.1
openslide-python==1.1.0
matplotlib==2.2.2
tensorboardX
```

The version constraint **==** will introduce the risk of dependency conflicts because the scope of dependencies is too strict.
The version constraint **No Upper Bound** and **\*** will introduce the risk of the missing API Error because the latest version of the dependencies may remove some APIs.

After further analysis, in this project,
The version constraint of dependency **scipy** can be changed to *>=0.12.0,<=1.7.3*.
The version constraint of dependency **pillow** can be changed to *==9.2.0*.
The version constraint of dependency **pillow** can be changed to *>=2.0.0,<=9.1.1*.
The version constraint of dependency **scikit-image** can be changed to *>=0.13.0,<=0.18.3*.

The above modification suggestions can reduce the dependency conflicts as much as possible, 
and introduce the latest version as much as possible without calling Error in the projects.

The invocation of the current project includes all the following methods.

<details><summary>The calling methods from the scipy</summary>
 <pre>scipy.ndimage.morphology.binary_fill_holes
scipy.ndimage.distance_transform_edt
</pre>
</details>
<details><summary>The calling methods from the pillow</summary>
 <pre>PIL.Image.open
</pre>
</details>
<details><summary>The calling methods from the scikit-image</summary>
 <pre>skimage.filters.threshold_otsu
skimage.measure.points_in_poly
</pre>
</details>
<details><summary>The calling methods from the all methods</summary>
 <pre>m.bias.data.zero_
x.get
computeITCList
pid.self._annotations.inside_polygons
torch.bmm
self.layer4
set
torch.load
scipy.ndimage.distance_transform_edt
skimage.measure.points_in_poly
m.weight.data.normal_
time.strftime
FP_probs.append
args.patch_size.args.patch_size.args.level.y.x.slide.read_region.convert.save
self.layer3
Probs.append
numpy.float32.img.np.array.transpose.transpose
self._resolution.np.log2.is_integer
torch.norm
numpy.random.seed
i.positive_vertices.astype
self.bn2
numpy.transpose
numpy.where
open
compute_FP_TP_Probs
os.mkdir
numpy.zeros
numpy.float32.img.np.array.transpose
self.conv1
computeEvaluationMask
torch.nn.Sequential
logging.basicConfig
numpy.amax
line.strip.split
Thresh.unlisted_FPs.np.asarray.sum
matplotlib.pyplot.plot
args.level.slide.level_dimensions.args.level.slide.read_region.convert
matplotlib.pyplot.show
numpy.rot90
skimage.measure.label
Xcorr.append
numpy.round
copy.deepcopy
range
self.bn3
matplotlib.pyplot.xlabel
os.path.join
torch.nn.Linear
numpy.random.randint
numpy.asarray.append
wsi.data.annotation.Formatter.camelyon16xml2json
int.strip
numpy.interp
torch.nn.Parameter
json_dict.append
logits.sigmoid
skimage.filters.gaussian.max
opts_list.append
ckpt.cpu.numpy
self._preprocess
xml.etree.ElementTree.parse
self.fc
self.Bottleneck.super.__init__
numpy.random.rand
str
numpy.asarray
numpy.logical_not
self.maxpool
skimage.measure.regionprops
matplotlib.pyplot.figure
int.Y.X.np.round.astype.transpose
torch.nn.ReLU
self.avgpool
sys.path.append
math.sqrt
each.endswith
wsi.data.annotation.Annotation.from_json
self.ResNet.super.__init__
argparse.ArgumentParser
print
openslide.open_slide.read_region
open.write
inxml.ET.parse.getroot.findall
matplotlib.pyplot.imshow
multiprocessing.Pool.map
self._image_size.self._image_size.y.x.self._slide.read_region.convert
run
os.path.dirname
feats.view.view
Polygon.inside
skimage.color.rgb2hsv
float
sorted
self.downsample
format
torch.nn.MaxPool2d
list
int.Y.X.np.round.astype.transpose.tolist
torch.zeros
numpy.max
argparse.ArgumentParser.parse_args
self.BasicBlock.super.__init__
skimage.filters.gaussian
ckpt.cpu.numpy.reshape
torch.nn.Conv2d
i.negative_vertices.astype
openslide.OpenSlide.read_region
shutil.copyfile
csvDIR.open.readlines
multiprocessing.Pool
int.i.negative_vertices.astype.tolist
self._polygons_negative.append
os.path.abspath
numpy.fliplr
plotFROC
logging.info
inxml.ET.parse.getroot
Y.X.np.round.astype
matplotlib.pyplot.subplot
self._color_jitter
numpy.log2
self.layer1
self.crf
self.layer2
torch.nn.BatchNorm2d
Isolated_Tumor_Cells.append
torch.transpose
numpy.mean
sum
computeFROC
line.rstrip
self.CRF.super.__init__
args.patch_size.args.patch_size.args.level.y.x.slide.read_region.convert
sys.stdout.flush
len
self._make_layer
map
torchvision.transforms.ColorJitter
wsi.model.layers.CRF
numpy.load
argparse.ArgumentParser.add_argument
matplotlib.pyplot.figure.suptitle
Ycorr.append
line.strip
conv3x3
block
layers.append
self._slide.read_region
pow
Thresh.unlisted_TPs.np.asarray.sum
super
torch.sum
matplotlib.pyplot.ylabel
logits.clone
int
self.avgpool.size
readCSVContent
enumerate
os.path.exists
x.vertices
multiprocessing.Value
self._coords.append
self.bn1
PIL.Image.open
skimage.filters.threshold_otsu
line.rstrip.split
m.weight.data.fill_
numpy.save
self.avgpool.view
self.relu
total_TPs.append
json.load
self._polygons_positive.append
numpy.float32.img.np.array.transpose.rotate
ResNet
os.listdir
main
json.dump
numpy.array
annotation.findall
self.conv2
self.modules
torch.squeeze
scipy.ndimage.morphology.binary_fill_holes
self.conv3
torch.nn.AvgPool2d
openslide.OpenSlide
multiprocessing.Lock
torch.squeeze.view
torch.load.cpu
open.close
wsi.data.annotation.Annotation
Exception
Polygon
int.i.positive_vertices.astype.tolist
openslide.open_slide
isinstance
</pre>
</details>

@developer
Could please help me check this issue?
May I pull a request to fix it?
Thank you very much.
